lambdas/nino-check-function/src/nino-check-handler.ts

```diff
@@ -0,0 +1,533 @@
+      body,
+      sessionItem.expiryDate,
+      "PASS"
+    );
+    return true;
+  } else if (
+    matchingResponse.status === 401 ||
+    matchingResponse.status === 424
+  ) {
+    logger.info("Storing user attempt as FAIL");
+    await storeUserAttempt(
+      sessionItem.clientSessionId,
+      matchingResponse.status.toString(),
+      body,
+      sessionItem.expiryDate,
+      "FAIL"
+    );
+    return false;
+  } else if (json.code === "INVALID_CREDENTIALS") {
+    // perhaps call OTG again and re-try
+    throw new Error("Matching API responded with INVALID_CREDENTIALS");
+  } else {
+    // retry 3x before throwing?
+    throw new Error(
+      "Matching API responded with unexpected status " + matchingResponse.status
+    );
+  }
+}
+
+async function storeUserAttempt(
+  sessionId: string,
+  status: string,
+  body: string,
+  expiryDate: string,
+  attempt: string
+) {
+  logger.info("Storing user attempt into database");
+
+  await dynamoClient.send(
+    new PutItemCommand({
+      TableName: process.env.UserAttemptsTable,
+      Item: {
+        sessionId: {
+          S: sessionId,
+        },
+        timestamp: {
+          S: Math.floor(Date.now() / 1000).toString(),
+        },
+        status: {
+          S: status,
+        },
+        text: {
+          S: body,
+        },
+        attempt: {
+          S: attempt,
+        },
+        ttl: {
+          N: expiryDate,
+        },
+      },
+    })
+  );
+}
+
+async function storeAuthorizationCode(
+  sessionTableName: string,
+  sessionId: string
+) {
+  logger.info("Storing auth code and expiry");
+
+  await dynamoClient.send(
+    new UpdateItemCommand({
+      TableName: sessionTableName,
+      Key: {
+        sessionId: {
+          S: sessionId,
+        },
+      },
+      UpdateExpression:
+        "SET authorizationCode = :authCode, authorizationCodeExpiryDate = :expiry",
+      ExpressionAttributeValues: {
+        ":authCode": {
+          S: randomUUID(),
+        },
+        ":expiry": {
+          N: (Math.floor(Date.now() / 1000) + 600).toString(),
+        },
+      },
+    })
+  );
+}
+
+async function saveNinoToUserAttemptsTable(
+  sessionId: string,
+  nino: string,
+  expiryDate: string
+) {
+  await dynamoClient.send(
+    new PutItemCommand({
+      TableName: process.env.NinoUsersTable,
+      Item: {
+        sessionId: {
+          S: sessionId,
+        },
+        nino: {
+          S: nino,
+        },
+        ttl: {
+          S: expiryDate,
+        },
+      },
+    })
+  );
+}
+
+function http400() {
+  return {
+    statusCode: 400,
+    body: "",
+  };
+}
+
+function http500() {
+  return {
+    statusCode: 500,
+    body: "",
+  };
+}
+
+function http200(requestRetry: boolean) {
+  return {
+    statusCode: 200,
+    body: JSON.stringify({
+      requestRetry: requestRetry,
+    }),
+  };
+}
+
+const handlerClass = new NinoCheckHandler();
+export const lambdaHandler = handlerClass.handler.bind(handlerClass);
```

Comment:
Ok yeah so this lambda is quite big, but I guess once we "productionise" the code we'd probably be looking at more like 200 lines? There's nothing that feels suuper complex